src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+        }
+
+        public Task<ImmutableArray<EditorCompletion.CompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            session.ItemCommitted += ItemCommitted;
+            session.Dismissed += SessionDismissed;
+            return Task.FromResult(data.InitialList.OrderBy(i => i.SortText).ToImmutableArray());
+        }
+
+        private void ItemCommitted(object sender, EditorCompletion.CompletionItemEventArgs e)
+        {
+            MakeMostRecentItem(e.Item.DisplayText);
+        }
+
+        private void SessionDismissed(object sender, EventArgs e)
+        {
+            if (sender is IAsyncCompletionSession session)
+            {
+                session.ItemCommitted -= ItemCommitted;
+                session.Dismissed -= SessionDismissed;
+            }
+        }
+
+        public Task<EditorCompletion.FilteredCompletionModel> UpdateCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            return Task.FromResult(UpdateCompletionList(session, data, cancellationToken));
+        }
+
+        public EditorCompletion.FilteredCompletionModel UpdateCompletionList(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            if (!session.Properties.TryGetProperty<bool>(CompletionItemSource.MustSetSelection, out bool mustSetSelection))
+            {
+                // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                // For now, the default mustSetSelection is false. Can discuss if the opposite is required.
+                mustSetSelection = false;
+            }
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+
+            // Check if the user is typing a number. If so, only proceed if it's a number
+            // directly after a <dot>. That's because it is actually reasonable for completion
+            // to be brought up after a <dot> and for the user to want to filter completion
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
```

Comment:
can you point to the original place this code exists.  i want to check to make sure the logic is preserved, and it's done at hte approprite layer across all languages.  thanks!

Reply:
[http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs,117](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs,117)

---
In reply to: [222085787](https://github.com/dotnet/roslyn/pull/29016#discussion_r222085787) [](ancestors = 222085787)